bpf/bpf_lxc.c

```diff
@@ -1243,6 +1262,12 @@ int tail_ipv4_policy(struct __ctx_buff *ctx)
 
 	/* Store meta: essential for proxy ingress, see bpf_host.c */
 	ctx_store_meta(ctx, CB_PROXY_MAGIC, ctx->mark);
+
+#ifdef ENABLE_CUSTOM_CALLS
+	encode_custom_prog_meta(ctx, ret, src_label, CUSTOM_CALLS_IDX_INGRESS);
+	tail_call_static(ctx, &CUSTOM_CALLS_MAP, CUSTOM_CALLS_IDX_INGRESS);
+#endif
```

Comment:
I note that this seems to run on egress prior to reaching the egress proxy, and on ingress prior to reaching the ingress proxy. This doesn't seem entirely balanced. Do you have a model in mind of how this should interact with L7 proxies?

Reply:
The placement of the tail call hooks was mostly dictated by the constraints coming with the hooks: I run the custom program at the very last stage, as I cannot “jump back” to the datapath code.

I probably didn't give as much attention to the L7 proxy as I should have, so no, I don't really have a model in mind at the moment.

Reading through the code, it looks like the decision to redirect to the proxy is taken in `ipv4_policy()`, but is not achieved before we return from `tail_ipv4_policy()`, _after_ the tail call hook. So I'm wondering... Maybe I'm actually running the custom program on each trip to and from the proxy? I need to double-check and run tests. If that's the case I can add a restriction on `ret == POLICY_ACT_PROXY_REDIRECT` I suppose. Maybe I can find a way to run _after_ L7 proxy on ingress to get something more balanced?

What information exactly would we loose if the hooks remain unbalanced?

Reply:
I did have a double processing, before and after the packet was processed by the ingress proxy. Addressed with a guard on `ret != CTX_ACT_TX`.